### ISSUE TITLE
Emitter Fault

### ISSUE THREAD

**Issue by killjoy7** [ParadiseSS13/Paradise#9095]
Odd one for you all, can't seem to find any entries indicating this has been happening to others.

Just reopened my server for the first time in a while, downloaded a fresh version of parastation. Pretty much unedited except for the usual config setup and baseline edits necessary to get it up and running. 

The oddity is that the four emitters in engineering flat-out refuse to operate about 9/10 times. They'll turn on (i.e. they light up), sufficient power will be in the grid, checked with multitool, even reset the powernets just to be sure. But they'll never fire. No variables in the local entities appear to be set apart from the default other than direction. There does seem to be some oddity in the default variables which i will get to in a moment. Then again i have no baseline to compare them against so i really don't know if that's an issue or not. Also, on rare occasion, they'll just work from the start of the round with no explanation or different approach.

I have found something of a workaround but it's hardly piratical. Using the tweak component ratings  admin command (which, admittedly, I don't fully understand the function of) on the emitter while in game and setting the value to 100 seems to immediately resolve the issue. Upon checking the entity i get the following changes:

[Value on left is the working emitter, post tweak. Value on right is default, not working]
active_power_usage == -4650 || 300 
fire_delay == 2 || 100
in_use == 1 || 0 
maximum_fire_delay == -1880 || 100 
minimum_fire_delay = -376 || 20
powered = 1 || 0

After some trial and error, i've determined that the only variable that actually needs to be changed in order for the emitter to function is the active_power_usage variable. All of the others do not seem to have any major impact one way or another. 

The other oddity is that I've attempted to edit the local entities directly in Dream Maker to force the variable to default to the new working value and even changed the code found within code\modules\power\singularity\emitter.dm directly to modify the default from 300 to -4650 just to see if i could get a reliable startup. However, I've since noticed that for reasons I can't quite grasp, no matter what i seem to change in Dream Maker (yes, i've been recompiling each time as well), the active_power_usage variable on those four emitters refuses to start at anything other than 300 and thus I can't even get the workaround to function. My guess is that there's some other area-based function that's adjusting this variable that I can't locate. 

I'm no coder, and i have a tendency to just bash things until they work, so the odds of there being some incredibly simple solution i'm overlooking is high. But all the same, any advice on either how to fix this or another approach I should attempt would be greatly appreciated.

TLDR:
Emitters appear to turn on but don't fire.
Snag seems to be within the active_power_usage variable
Changing the variable to negate power consumption fixes the problem
Adjusting the code to make that change permanent does not seem to effect the emitters in-game.

Thoughts?

**Comment by killjoy7** [comment]
Update with workaround:

I located the following section of code within the code\modules\power\singularity\emitter.dm section and adjusted var\power_usage to -4650, the same value that tweak component ratings sets this value to in-game. The reflected value in active_power_usage in the emitters around engineering is now -4700 by default (note the equation towards the end of the code below). 

/obj/machinery/power/emitter/RefreshParts()
	var/max_firedelay = 120
	var/firedelay = 120
	var/min_firedelay = 24
	var/power_usage = 350
	for(var/obj/item/stock_parts/micro_laser/L in component_parts)
		max_firedelay -= 20 * L.rating
		min_firedelay -= 4 * L.rating
		firedelay -= 20 * L.rating
	maximum_fire_delay = max_firedelay
	minimum_fire_delay = min_firedelay
	fire_delay = firedelay
	for(var/obj/item/stock_parts/manipulator/M in component_parts)
		power_usage -= 50 * M.rating
	active_power_usage = power_usage

I'm mostly just listing this as a quick fix for anyone that may be having the same issue. Though given this finding i'm now reasonably sure that the error actually stems from something having to do with the internal components of the emitters themselves. I don't fully understand the math in the code here or where everything is coming/going but i'm thinking that the components aren't being loaded properly and the resultant math is producing a number that isn't cooperating with the power grid. As far as i can tell, the only major downside of this workaround is that i'm pretty sure emitters will probably now work without their parts and that when all four are on it actually supplies something like ~18k watts to the power grid, though this could probably be negated by just adjusting the number down to -1 or something along those lines.

Anyway, i'll keep playing with it and update if i find a true fix. Please add a comment if you find a less barbaric fix to this issue!

**Comment by Citinited** [comment]
Maybe the SMES were drawing more power than was available in the grid?

**Comment by alexkar598** [comment]
Deconstruct the emoter by using welding tool,wrench,screwdriver,crowbar and rebuild it

**Comment by killjoy7** [comment]
@Citinited As far as i can recall, the emitters in engineering operate on a separate grid maintained by the single SMES off to the left, additionally i'm pretty sure that the three on the right aren't set to charge by default. Good thought tho, i'll double check that in future startups to make sure that's not the issue. 

@alexkar598 I'll give that a try later with the default settings re-enabled to see if that corrects their function for the duration of the shift. However, that's not really a realistic fix as the engineers shouldn't have to disassemble and reassemble the emitters at the start of the round every time.

**Comment by Kyep** [comment]
Apparently this is a known bug - emitters often don't turn on at all unless you destroy and rebuild them first.

**Comment by MINIMAN10000** [comment]
The emitters are the first thing added to the powernet making them the last thing in the powernet. When you are rebuilding the emitter you are making it the last thing in the power net and thus the first thing to get power. Each one requires 300 power as stated. This means the SMES that power the main grid get power first, and thus eat all the power before it reaches the emitter.

This means in order to solve this problem a map edit needs to be done to put the emitters as the last things in the power net and that should solve it.
